nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -0,0 +1,194 @@
+    # solve issues:
+    # - one package incoming, one package outgoing (synchronize)
+    # - mark buffer if new package arrived --> not to slow not to fast 
+    # - FIRST idea: 
+    #   - Receiving rank 0 check for 'NaN' in buffer, i.e. is it empty?
+    #   - if NaN then fill with new package
+    #   - Sending rank 1 (or more in the future) check for new package in buffer (not NaN)
+    #   - do science and send to TVB
+    #   - set buffer to NaN to mark as ready
+    # TODO: how to check for END of simulation?
+    
+    ############ OLD Code to replace: with threads, current solution
+    initialisation =np.load(param['init']) # initialisation value
+    buffer=[initialisation] # buffer of the rate to send it
+        
+    # create the thread for receive and send data
+    th_receive = Thread(target=receive, args=(logger_receive,store,status_data,buffer, comm_receiver))
+    th_send = Thread(target=send, args=(logger_send,analyse,status_data,buffer, comm_sender))
+
+    # start the threads
+    # FAT END POINT
+    logger_master.info('Start thread')
+    th_receive.start()
+    th_send.start()
+    th_receive.join()
+    th_send.join()
+    logger_master.info('thread join')
+    ############# OLD Code end
+
+
+def _shared_mem_buffer(buf_pkgs,sizes,comm):
+    '''
+    Create a shared memory block.
+    MPI One-sided-Communication.
+    NOTE: still copy-paste from kim master thesis code...
+    '''
+    pkg_size = np.sum(sizes)
+    datasize = MPI.DOUBLE.Get_size()
+    if comm.Get_rank() == 0:
+        bufbytes = buf_pkgs * datasize*2 * pkg_size 
+    else: 
+        bufbytes= 0
+    # rank 0: create the shared block
+    # rank 1-x: get a handle to it
+    win = MPI.Win.Allocate_shared(bufbytes, datasize, comm=comm)
+    buf, datasize = win.Shared_query(0) 
+    assert datasize == MPI.DOUBLE.Get_size()
+    # create a numpy array (buffer) whose data points to the shared mem
+    return np.ndarray(buffer=buf, dtype='d', shape=(buf_pkgs*pkg_size,2))
+
+
+
+def receive(logger,store,status_data,buffer, comm):
+    '''
+    the receive part of the translator
+    :param logger : logger fro the thread
+    :param store : object for store the data before analysis
+    :param status_data: the status of the buffer (SHARED between thread)
+    :param buffer: the buffer which contains the data (SHARED between thread)
+    :return:
+    '''
+    # initialise variables for the loop
+    status_ = MPI.Status() # status of the different message
+    source_sending = np.arange(0,comm.Get_remote_size(),1) # list of all the process for the commmunication
+    check = np.empty(1,dtype='b')
+    count=0
+    while True: # FAT END POINT
+        # send the confirmation of the process can send data
+        logger.info(" Nest to TVB : wait all")
+        for source in source_sending:
+            comm.Recv([check, 1, MPI.CXX_BOOL], source=source, tag=MPI.ANY_TAG, status=status_)
+
+        if status_.Get_tag() == 0:
```

Comment:
what are different tags in status?

Reply:
Added this to [Issue 3](https://github.com/multiscale-cosim/TVB-NEST/issues/3)